### PR TITLE
Log file client development

### PR DESCRIPTION
**Pull request by anna-st-40**
Implemented the base functionality of the log file client app. 
It can now read CSV logs and display them as a table of the first 100 lines (displaying all the lines takes too long to load).

### CODE REVIEW

File: extras/log_file_client/lib/utils/csv_reader.dart [hunk 1]
```diff
@@ -0,0 +1,109 @@
+import 'dart:async';
+import 'dart:io';
+
+import 'package:flutter/services.dart';
+import 'package:http/http.dart' as http;
+
+abstract class CsvReader {
```
Comment (jgfoster) [inline extras/log_file_client/lib/utils/csv_reader.dart:7]: Nice job of abstract class use.

File: extras/log_file_client/lib/utils/log_list.dart [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+}
+
+class LogListReaderForTest implements LogListReader {
+  @override
+  Future<List<Log>> fetchList() async {
+    return [Log('test1', '/test1.test'), Log('test2', '/test2.test'), Log('test3','/test3.test')];
+  }
+}
+
+class LogListReaderForAppWeb implements LogListReader {
```
Comment (jgfoster) [inline extras/log_file_client/lib/utils/log_list.dart:22]: This has a lot of shared code with the next class. Perhaps you could create a method in the abstract superclass that takes HTML and returns the `List<Log>`. In fact, you could modify the `fetchList` above to have a hard-coded HTML "file" that tests the new method.

File: extras/log_file_client/test/csv_reader_test.dart [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+
+    expect(newCsvReaderTest.filePath, equals('test1.csv'));
+    expect(newCsvReaderAppWeb.filePath, equals('test2.csv'));
+    expect(newCsvReaderAppLocal.filePath, equals('test3.csv'));
+  });
+
+  test_package.test('fetchCsvData with test class', () async {
+    final reader = CsvReaderForTest('test/test_assets/csv_reader_test.csv');
+    final data = await reader.fetchCsvData();
+    expect(data, 'time,tankid,temp,temp setpoint,pH,pH setpoint,onTime,Kp,Ki,Kd\r\n01/20/2023 16:18:21,  99, 0.00, 10.00, 0.000, 8.645,    6,    700.0,    100.0,      0.0\r\n01/20/2023 16:18:22,  98, 1.23, 10.00, 7.123, 8.645,    8,    710.0,    110.0,      1.0\r\n01/20/2023 16:18:23,  97, 2.34, 10.00, 6.789, 8.645,    9,    720.0,    120.0,      2.0\r\n01/20/2023 16:18:24,  96, 3.45, 10.00, 5.456, 8.645,   10,    730.0,    130.0,      3.0\r\n01/20/2023 16:18:25,  95, 4.56, 10.00, 4.123, 8.645,   11,    740.0,    140.0,      4.0\r\n');
```
Comment (jgfoster) [inline extras/log_file_client/test/csv_reader_test.dart:23]: This is probably more clear as to what is expected, but for a really long string you could get an [MD5](https://stackoverflow.com/questions/47870642/dart-md5-from-string) and compare that.

**Review by jgfoster — CHANGES_REQUESTED** [review]
This is very impressive. I've added comments and would like to discuss alternatives to the CORS issue.

File: extras/log_file_client/lib/components/csv_view.dart [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+          return const Center(child: CircularProgressIndicator());
+        } else if (snapshot.hasError) {
+          return Center(child: Text('Error: ${snapshot.error}'));
+        } else if (!snapshot.hasData || snapshot.data!.isEmpty) {
+          return const Center(child: Text('No data found'));
+        } else {
+          final csvTable = snapshot.data!;
+          return SingleChildScrollView(
+            scrollDirection: Axis.horizontal,
+            child: SingleChildScrollView(
```
[the remark this answers is not in the document]
Comment (anna-st-40) [reply extras/log_file_client/lib/components/csv_view.dart:36]: Changed it. I didn't realize that ListView optimizes by only rendering the widgets in view. This got rid of the slow loading time issue.

File: extras/log_file_client/lib/pages/home_page.dart [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+  final LogListReader? logListReader;
+
+  @override
+  State<HomePage> createState() => _HomePageState();
+}
+
+class _HomePageState extends State<HomePage> {
+  late final LogListReader logListReader;
+  List<Log>? _logList;
+  bool _isLoading = true;
```
[the remark this answers is not in the document]
Comment (anna-st-40) [reply extras/log_file_client/lib/pages/home_page.dart:18]: This doesn't work and I'm not entirely sure why. It causes indefinite loading.

File: extras/log_file_client/lib/utils/csv_reader.dart [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+    } else {
+      throw Exception('Unable to format date string $dateString.');
+    }
+  }
+
+  Future<List<List>> csvTable() {
+    return fetchCsvData().then((String data) {
+      final List<List> table = [];
+
+      final List<String> rows = data.split('\n');
```
[the remark this answers is not in the document]
Comment (anna-st-40) [reply extras/log_file_client/lib/utils/csv_reader.dart:27]: You're right. Much easier!

File: extras/log_file_client/test/csv_reader_test.dart (hunk 3 above)
Comment (anna-st-40) [reply extras/log_file_client/test/csv_reader_test.dart:23]: I found leaving it as a long string makes the test more useful to see what exactly went wrong if it fails.

File: extras/log_file_client/lib/pages/home_page.dart (hunk 5 above)
[the remark this answers is not in the document]
Comment (jgfoster) [reply extras/log_file_client/lib/pages/home_page.dart:18]: As mentioned, there are trade-offs, and while this is fine I'd like to understand why the other approach didn't work. Let's try to remember to look at this when we meet.

**Review by jgfoster — APPROVED** [review]

**Pull request merged**